db-engine/src/main/java/com/splicemachine/db/impl/sql/execute/ValueRow.java

```diff
@@ -400,6 +411,8 @@ public void readExternal(ObjectInput in) throws IOException, ClassNotFoundExcept
 		for (int i = 0; i < ncols; i++) {
 			column[i] = (DataValueDescriptor) in.readObject();
 		}
+		StructType structType = (StructType)in.readObject();
```

Comment:
I feel that serializing the schema for every single ValueRow is too much overhead

Reply:
The schema was added so that the dataframe returned from SplicemachineContext.df() would contain a schema having the correct column names instead of the C0, C1, etc, names generated by ValueRow.getNamedColumn().  Since individual rows are returned through Kafka rather than one collection object of rows, I put the schema on the row level.  org.apache.spark.sql.Row implemented by ValueRow also has a method for returning the schema on the row level.

Would it be acceptable if ValueRow stores and serializes the schema as a string, and converts it to StructType only when it’s needed in ValueRow.schema() ?

Reply:
I would try to get the schema separately, maybe through the JDBC connection, and feed it to KafkaToDF() so that it has it from the beginning rather than expecting it on every ValueRow

Reply:
Fixed in new commit [5eb097b](https://github.com/splicemachine/spliceengine/pull/3391/commits/5eb097b492ce989c69231349fcecfa88c5554952).